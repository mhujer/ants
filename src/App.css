.game {
    display: flex;
    flex-direction: row;
}

.playerDashboard {
    flex: 2;
    opacity: 0.5;
    max-width: 15em;
}

.playerDashboard.is-on-turn {
    opacity: 1;
    border: 3px seagreen solid;
}

.playerStats,
.castleStats {
    border: 1px dotted black;
    margin: 3px;
}

.not-playable {
    opacity: 0.4;
}
