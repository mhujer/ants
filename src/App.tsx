import React, { useState } from 'react';
import { Player } from './Player';

export type Players = 'player1' | 'player2';

const App: React.FC = () => {
    const initialPlayer: Players = 'player1';

    const [currentPlayer, setCurrentPlayer] = useState<Players>(initialPlayer);
    const [player1WallHeight, setPlayer1WallHeight] = useState(20);
    const [player2WallHeight, setPlayer2WallHeight] = useState(20);

    const playHandler = () => {
        if (currentPlayer === 'player1') {
            setPlayer1WallHeight((wallHeight) => wallHeight + 1);
            setCurrentPlayer('player2');
        } else {
            setPlayer2WallHeight((wallHeight) => wallHeight + 2);
            setCurrentPlayer('player1');
        }
    };

    return (
        <>
            <Player
                player="player1"
                currentPlayer={currentPlayer}
                playHandler={playHandler}
                wallHeight={player1WallHeight}
            />
            <Player
                player="player2"
                currentPlayer={currentPlayer}
                playHandler={playHandler}
                wallHeight={player2WallHeight}
            />
        </>
    );
};

export default App;
