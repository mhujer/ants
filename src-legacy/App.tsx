import React, { useEffect, useRef } from 'react';
import { PlayerDashboard } from './PlayerDashboard';
import { Castle } from './Castle';
import './App.css';
import { CardComponent } from './CardComponent';
import { Card } from './Card';
import { Player } from './Player';
import { Cards } from './Cards';
import { playSound, Sound } from './Sounds';
import { useAppDispatch, useAppSelector } from './store/hooks';
import { discardCard, playCard, playSound as playSoundAction, selectGame } from './store/game/gameSlice';
import { cumulativeOffset } from '../src/utils';

const App: React.FC = () => {
    const gameState = useAppSelector(selectGame);
    const dispatch = useAppDispatch();

    const discardDeckWrapperRef = useRef<HTMLDivElement>(null);

    /*
    const playerOnTurn = useAppSelector((state) => state.game.playerOnTurn);

    useEffect(() => {
        const timeoutId = setTimeout(() => {
            // @todo canCardBePlayed asi přehodit do state
            const aiAction = dummyAi(gameState);
            console.dir('AI...');
            console.dir(aiAction);
            if (aiAction.type === 'play') {
                dispatch(playCard(aiAction.card));
            } else if (aiAction.type === 'discard') {
                dispatch(discardCard(aiAction.card));
            }
        }, 200);

        return () => {
            clearTimeout(timeoutId);
        }
    }, [playerOnTurn, gameState]);*/

    // play sound
    useEffect(() => {
        setTimeout(() => {
            if (gameState.playSound !== null) {
                playSound(gameState.playSound);
                dispatch(playSoundAction());
                // deal new card
                //setTimeout(() => {
                //    playSound(Sound.CARD_PLAYED);
                //}, 500);
            }
        }, 500);
    }, [gameState.playSound]);

    // play card sound
    useEffect(() => {
        if (gameState.lastPlayedCard !== undefined) {
            playSound(Sound.CARD_PLAYED);
        }
    }, [gameState.lastPlayedCard]);

    function playCardHandler(card: Card) {
        dispatch(playCard(card));
    }

    function discardCardHandler(card: Card) {
        dispatch(discardCard(card));
    }

    let playerWonContent = null;
    if (gameState.playerWon !== undefined) {
        if (gameState.playerWon === Player.BLACK_ANTS) {
            playerWonContent = <h1>Černí mravenci vyhráli!</h1>;
        }
        if (gameState.playerWon === Player.RED_ANTS) {
            playerWonContent = <h1>Červení mravenci vyhráli!</h1>;
        }
    }

    const playerOnTurnData = gameState.playerOnTurn === Player.BLACK_ANTS ? gameState.playerBlack : gameState.playerRed;

    //const allCards = cardDefinitions.map((cardDefinition) => new Card(cardDefinition));
    //const allCardsComponents = allCards.map((card) => <CardComponent card={card} key={card.id} />);

    let discardDeckWrapperX = 0;
    let discardDeckWrapperY = 0;
    if (discardDeckWrapperRef.current !== null) {
        //discardDeckWrapperX = discardDeckWrapperRef.current.offsetLeft;
        //discardDeckWrapperY = discardDeckWrapperRef.current.offsetTop;
        const offset = cumulativeOffset(discardDeckWrapperRef.current);
        discardDeckWrapperX = offset.left;
        discardDeckWrapperY = offset.top;
    }

    return (
        <>
            {playerWonContent}
            <div className="game">
                <PlayerDashboard
                    isOnTurn={gameState.playerOnTurn === Player.BLACK_ANTS}
                    builders={gameState.playerBlack.builders}
                    bricks={gameState.playerBlack.bricks}
                    soldiers={gameState.playerBlack.soldiers}
                    weapons={gameState.playerBlack.weapons}
                    mages={gameState.playerBlack.mages}
                    crystals={gameState.playerBlack.crystals}
                    castle={gameState.playerBlack.castle}
                    wall={gameState.playerBlack.wall}
                />
                <Castle castle={gameState.playerBlack.castle} wall={gameState.playerBlack.wall} />
                <div ref={discardDeckWrapperRef}>
                    {gameState.lastPlayedCard !== undefined ? (
                        <CardComponent
                            key={gameState.lastPlayedCard.id}
                            card={gameState.lastPlayedCard}
                            coords={{ x: 0, y: 0 }}
                        />
                    ) : null}
                    {/*<CardComponent key={gameState.playerBlack.cards[0]!.id} card={gameState.playerBlack.cards[0]} />*/}
                </div>
                <Castle castle={gameState.playerRed.castle} wall={gameState.playerRed.wall} />
                <PlayerDashboard
                    isOnTurn={gameState.playerOnTurn === Player.RED_ANTS}
                    builders={gameState.playerRed.builders}
                    bricks={gameState.playerRed.bricks}
                    soldiers={gameState.playerRed.soldiers}
                    weapons={gameState.playerRed.weapons}
                    mages={gameState.playerRed.mages}
                    crystals={gameState.playerRed.crystals}
                    castle={gameState.playerRed.castle}
                    wall={gameState.playerRed.wall}
                />
            </div>
            <Cards
                discardDeckCoordinates={{ x: discardDeckWrapperX, y: discardDeckWrapperY }}
                playerResources={{
                    bricks: playerOnTurnData.bricks,
                    weapons: playerOnTurnData.weapons,
                    crystals: playerOnTurnData.crystals,
                }}
                cardsInHand={playerOnTurnData.cards}
                playCardHandler={(card: Card) => playCardHandler(card)}
                discardCardHandler={(card: Card) => discardCardHandler(card)}
            />
            {/*<div style={{ display: 'flex', flexDirection: 'row' }}>{allCardsComponents}</div>*/}
        </>
    );
};

export default App;
