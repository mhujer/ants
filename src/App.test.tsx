import App from './App';
import renderer from 'react-test-renderer';
import { describe, expect, it } from 'vitest';

describe('App', () => {
    it('renders timer for 0 seconds', () => {
        const domTree = renderer.create(<App />).toJSON();
        expect(domTree).toMatchSnapshot();
    });
});
