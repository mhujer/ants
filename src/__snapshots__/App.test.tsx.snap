// Vitest Snapshot v1

exports[`App > renders timer for 0 seconds 1`] = `
[
  <div
    style={
      {
        "border": "2px solid green",
      }
    }
  >
    <h1>
      20
    </h1>
    <button
      onClick={[Function]}
    >
      Build!
    </button>
  </div>,
  <div
    style={
      {
        "border": "",
      }
    }
  >
    <h1>
      20
    </h1>
  </div>,
]
`;
